### DISCUSSION TITLE
What is scope in Flax?

### DISCUSSION THREAD

**Discussion by vrvrv** [google/flax#1515] [Q&A] [answered]
Hello Flax! 

I have trouble with initializing model with pre-trained model. I'd like to train decoder-only generative model and I want to initialize the decoder(my new model) with pre-trained VAE model written by Flax. 

https://github.com/google/flax/blob/1a117d0aa0b9491d0abbd77e003ffe7de508cc49/examples/vae/train.py#L53-L89

However, I've stuck in some problem. I think this is simple problem but it is too hard for me to understand the underlying mechanisms. Could you help me? Here is my code base.

```python
class Almond(nn.Module):
    
    def setup(self):
        self.decoder = Decoder()
        
    def __call__(self, z):
        return self.decoder(z)

Almond().setup()

state = train_state.TrainState.create(
    params=pre_trained_vae_state.params['decoder'],
    tx=optax.adam(1e-3),
)
```

But when I run `Almond().setup()`, this yields following error message

```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-105-4661b26745dd> in <module>
----> 1 Almond().setup()

~/opt/anaconda3/envs/jax/lib/python3.6/site-packages/flax/linen/module.py in wrapped_module_method(*args, **kwargs)
    273     _context.module_stack.append(self)
    274     try:
--> 275       y = fun(self, *args, **kwargs)
    276       if _context.capture_stack:
    277         filter_fn = _context.capture_stack[-1]

<ipython-input-104-5f053e446753> in setup(self)
     12 
     13     def setup(self):
---> 14         self.decoder = Decoder()
     15 
     16     def __call__(self, z):

~/opt/anaconda3/envs/jax/lib/python3.6/site-packages/flax/linen/module.py in __setattr__(self, name, val)
    534     # Submodules are being defined and attached in setup()
    535     else:
--> 536       self._register_submodules(name, val)
    537 
    538   def __getattr__(self, name: str) -> Any:

~/opt/anaconda3/envs/jax/lib/python3.6/site-packages/flax/linen/module.py in _register_submodules(self, name, val)
    640 
    641   def _register_submodules(self, name, val):
--> 642     assert self.scope, 'Trying to register submodules on unbound scope.'
    643     root = self.scope.root
    644     cache = _caches.get(root, {})

AssertionError: Trying to register submodules on unbound scope.
```

I have a few questions about 
- what is submodules and scope saying here?
- Is it correct to start with `.setup()` rather thain `.init()`?

Thanks in advance!

**Comment by avital** [comment, marked as answer]
Hi @vrvrv, sorry -- the error message here should definitely be improved.

In general, Flax modules are stateless in top-level code. `setup()` is called once you instantiate a "bound module" which typically happens via `module.init` and `module.apply`. See https://flax.readthedocs.io/en/latest/flax.linen.html#flax.linen.Module.apply, or take a look at our MNIST example for a simple model with training loop and best practices

**Reply by shashank2000** [reply]
@avital I looked through the MNIST example and still don't quite understand what the initial poster seems to be doing wrong. How would one go about 'initializing' a pretrained model given that Flax modules are stateless in top-level code?

**Reply by marcvanzee** [reply]
The initial poster is manually calling `.setup()`, but you should never call this yourself. The Flax Module API uses `init()` and `apply()`.

If this is unclear to you, I recommend reading some of our "getting started" documentation:

- The [Flax basic tutorial](https://flax.readthedocs.io/en/latest/notebooks/flax_basics.html). 

- [Annotated MNIST](https://flax.readthedocs.io/en/latest/notebooks/annotated_mnist.html) example.

In the Annotated MNIST example you can see we initialize the params of a Module in `create_train_state` (Section 6) using `CNN().init()`, and we apply the params to inputs in `train_step` and `eval_step` usign `CNN().apply()` (Section 7 & 8).

Good luck!

**Reply by shashank2000** [reply]
perfect that makes sense thank you!
